# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [1]:
#load python packages
import os
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline

In [2]:
os.getcwd()

'/Users/xwang/Dropbox/Springboard/GuidedCapstone'

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [3]:
path = '/Users/xwang/Dropbox/Springboard/GuidedCapstone/data'
os.chdir(path)
dir_list = os.listdir()
print(dir_list)

['.DS_Store', 'step2_output.csv', 'step3_output.csv']


In [4]:
df = pd.read_csv('step3_output.csv')

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [5]:
dfo = df[['state']] 
df = pd.concat([df.drop(dfo, axis=1), pd.get_dummies(dfo)], axis=1)

In [6]:
df = df.drop(['Unnamed: 0', 'Unnamed: 0.1'], axis=1)
df.head(10)

Name  summit_elev  vertical_drop  trams  \
0               Hilltop Ski Area         2090            294      0   
1            Sunrise Park Resort        11100           1800      0   
2  Yosemite Ski & Snowboard Area         7800            600      0   
3         Boreal Mountain Resort         7700            500      0   
4                    Dodge Ridge         8200           1600      0   
5               Donner Ski Ranch         8012            750      0   
6                  June Mountain        10090           2590      0   
7            Mt. Shasta Ski Park         6890           1435      0   
8                  Mountain High         8200           1600      0   
9                 Ski China Peak         8709           1679      0   

   fastEight  fastSixes  fastQuads  quad  triple  double  ...  \
0        0.0          0          0     0       1       0  ...   
1        0.0          0          1     2       3       1  ...   
2        0.0          0          0     0       1       3  ...   
3        0.0          0          1     1       3       1  ...   
4        0.0          0          0     1       2       5  ...   
5        0.0          0          0     0       1       5  ...   
6        0.0          0          2     0       0       4  ...   
7        0.0          0          0     0       3       0  ...   
8        0.0          0          2     2       2       5  ...   
9        0.0          0          0     1       4       2  ...   

   state_Rhode Island  state_South Dakota  state_Tennessee  state_Utah  \
0                   0                   0                0           0   
1                   0                   0                0           0   
2                   0                   0                0           0   
3                   0                   0                0           0   
4                   0                   0                0           0   
5                   0                   0                0           0   
6                   0                   0                0           0   
7                   0                   0                0           0   
8                   0                   0                0           0   
9                   0                   0                0           0   

   state_Vermont  state_Virginia  state_Washington  state_West Virginia  \
0              0               0                 0                    0   
1              0               0                 0                    0   
2              0               0                 0                    0   
3              0               0                 0                    0   
4              0               0                 0                    0   
5              0               0                 0                    0   
6              0               0                 0                    0   
7              0               0                 0                    0   
8              0               0                 0                    0   
9              0               0                 0                    0   

   state_Wisconsin  state_Wyoming  
0                0              0  
1                0              0  
2                0              0  
3                0              0  
4                0              0  
5                0              0  
6                0              0  
7                0              0  
8                0              0  
9                0              0  

[10 rows x 60 columns]

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [7]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df.AdultWeekend 

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled = scaler.transform(X) 

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [8]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)


Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [9]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score, mean_absolute_error
lm = linear_model.Ridge(normalize = True)
model = lm.fit(X_train, y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [10]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)

## Review Model Outcomes — Iterate over additional models as needed

In [11]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model

In [12]:
print(explained_variance_score(y_test, y_pred))
print(mean_absolute_error(y_test, y_pred))

-0.05112831860784128
8.443016140379


**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [13]:
lm.intercept_

57.01097590790064

**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [14]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
dfcoef = pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])

In [15]:
dfcoef.sort_values(by=['Coefficient'], inplace=True, ascending=False)
print(dfcoef)

                      Coefficient
state_New Jersey         6.679256
AdultWeekday             3.677248
Runs                     1.574929
clusters                 1.459753
vertical_drop            1.435270
quad                     1.238057
triple                   1.088471
state_West Virginia      1.074514
fastQuads                1.015142
LongestRun_mi            0.975525
state_New Hampshire      0.972863
state_Idaho              0.904403
total_chairs             0.891210
TerrainParks             0.885793
state_Montana            0.863451
state_Virginia           0.830294
Snow Making_ac           0.792191
state_Maine              0.761142
state_Vermont            0.756420
state_North Carolina     0.749989
state_Colorado           0.662436
state_New Mexico         0.621059
state_Michigan           0.600750
state_Rhode Island       0.560458
state_Minnesota          0.530607
state_Oregon             0.504315
surface                  0.500116
state_Wyoming            0.492819
state_Ohio    

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [16]:
df = df.drop(df.filter(regex='^state',axis=1).columns, axis=1)

In [17]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df.AdultWeekend 

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled = scaler.transform(X) 

In [18]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

In [19]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)
y_pred = model.predict(X_test)

In [20]:
print(explained_variance_score(y_test, y_pred))
print(mean_absolute_error(y_test, y_pred))

0.5328005009266558
7.3715297405855615


In [21]:
lm.intercept_

56.35156951812893

In [22]:
dfcoef = pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])

In [23]:
dfcoef.sort_values(by=['Coefficient'], inplace=True, ascending=False)
print(dfcoef)

                    Coefficient
AdultWeekday       1.015202e+01
clusters           3.522739e+00
summit_elev        2.989118e+00
vertical_drop      2.686657e+00
TerrainParks       2.557659e+00
quad               2.053590e+00
daysOpenLastYear   1.819133e+00
projectedDaysOpen  1.692971e+00
SkiableTerrain_ac  1.516986e+00
surface            1.313251e+00
LongestRun_mi      1.270788e+00
Snow Making_ac     7.613298e-01
averageSnowfall    7.459696e-01
fastQuads          6.025261e-01
triple             5.147715e-01
double             5.146862e-01
total_chairs       4.763792e-01
yearsOpen          3.843848e-01
NightSkiing_ac     2.125304e-01
Runs               5.581900e-02
trams              8.881784e-16
fastSixes          8.881784e-16
fastEight          2.220446e-16


**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [24]:
df3 = df.drop(['summit_elev','vertical_drop'],axis=1)
df3.columns

Index(['Name', 'trams', 'fastEight', 'fastSixes', 'fastQuads', 'quad',
       'triple', 'double', 'surface', 'total_chairs', 'Runs', 'TerrainParks',
       'LongestRun_mi', 'SkiableTerrain_ac', 'Snow Making_ac',
       'daysOpenLastYear', 'yearsOpen', 'averageSnowfall', 'AdultWeekday',
       'AdultWeekend', 'projectedDaysOpen', 'NightSkiing_ac', 'clusters'],
      dtype='object')

In [25]:
# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df3.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df3.AdultWeekend 

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled = scaler.transform(X) 

In [26]:
# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)
y_pred = model.predict(X_test)

In [27]:
print(explained_variance_score(y_test, y_pred))
print(mean_absolute_error(y_test, y_pred))

0.5452369302341488
7.213858007800508


In [28]:
lm.intercept_

56.27484577237206

In [29]:
dfcoef = pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])
dfcoef.sort_values(by=['Coefficient'], inplace=True, ascending=False)
print(dfcoef)

                    Coefficient
AdultWeekday       1.044149e+01
clusters           3.374817e+00
TerrainParks       2.307123e+00
SkiableTerrain_ac  1.934719e+00
quad               1.852908e+00
LongestRun_mi      1.701832e+00
daysOpenLastYear   1.541736e+00
averageSnowfall    1.456394e+00
projectedDaysOpen  1.405688e+00
surface            1.257347e+00
Snow Making_ac     1.113826e+00
fastQuads          1.010383e+00
Runs               7.507877e-01
double             6.452499e-01
total_chairs       5.666666e-01
NightSkiing_ac     4.731105e-01
triple             4.109419e-01
yearsOpen          2.054270e-01
fastEight          6.661338e-16
fastSixes          4.440892e-16
trams              0.000000e+00


## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. |-0.05 | 8.44 |-|
| Model 2. | 0.53 | 7.37 |'state'|
| Model 3. | 0.55 | 7.21 |'state','summit_elev','base_elev'|

Model Selection:

In [ ]:
#Model3